src/rte/kernels/mo_optical_props_kernels.jl

```diff
@@ -43,53 +69,57 @@ contains
-      bind(C, name="delta_scale_2str_f_k")
-    integer,                               intent(in   ) :: ncol, nlay, ngpt
-    real(wp), dimension(ncol, nlay, ngpt), intent(inout) ::  tau, ssa, g
-    real(wp), dimension(ncol, nlay, ngpt), intent(in   ) ::  f
-
-    real(wp) :: wf
-    integer  :: icol, ilay, igpt
-
-    do igpt = 1, ngpt
-      do ilay = 1, nlay
-        do icol = 1, ncol
-          wf = ssa(icol,ilay,igpt) * f(icol,ilay,igpt)
-          tau(icol,ilay,igpt) = (1._wp - wf) * tau(icol,ilay,igpt)
-          ssa(icol,ilay,igpt) = (ssa(icol,ilay,igpt) - wf) /  max(eps,(1.0_wp - wf))
-          g  (icol,ilay,igpt) = (g  (icol,ilay,igpt) - f(icol,ilay,igpt)) / &
-                                        max(eps,(1._wp - f(icol,ilay,igpt)))
-        end do
-      end do
-    end do
-
-  end subroutine delta_scale_2str_f_k
+  function delta_scale_2str_f_k(ncol, nlay, ngpt, tau, ssa, g, f)
+#      bind(C, name="delta_scale_2str_f_k")
+#    integer,                               intent(in   ) :: ncol, nlay, ngpt
+#    real(wp), dimension(ncol, nlay, ngpt), intent(inout) ::  tau, ssa, g
+#    real(wp), dimension(ncol, nlay, ngpt), intent(in   ) ::  f
+
+#    real(wp) :: wf
+#    integer  :: icol, ilay, igpt
+
+    wp = eltype(wf)
+
+    for igpt = 1:ngpt
+      for ilay = 1:nlay
+        for icol = 1:ncol
+          wf = ssa[icol,ilay,igpt] * f[icol,ilay,igpt]
+          tau[icol,ilay,igpt] = (wp(1) - wf) * tau[icol,ilay,igpt]
+          ssa[icol,ilay,igpt] = (ssa[icol,ilay,igpt] - wf) /  max(ε_machine(wp),(wp(1) - wf))
+          g[icol,ilay,igpt] = (g[icol,ilay,igpt] - f[icol,ilay,igpt]) /
+                                        max(ε_machine(wp),(wp(1) - f[icol,ilay,igpt]))
+        end
+      end
+    end
+
+  end
   # ---------------------------------
   # Delta-scale
   #   f = g*g
   #
-  pure subroutine delta_scale_2str_k(ncol, nlay, ngpt, tau, ssa, g) &
-      bind(C, name="delta_scale_2str_k")
-    integer,                               intent(in   ) :: ncol, nlay, ngpt
-    real(wp), dimension(ncol, nlay, ngpt), intent(inout) ::  tau, ssa, g
-
-    real(wp) :: f, wf
-    integer  :: icol, ilay, igpt
-
-    do igpt = 1, ngpt
-      do ilay = 1, nlay
-        do icol = 1, ncol
-          f  = g  (icol,ilay,igpt) * g  (icol,ilay,igpt)
-          wf = ssa(icol,ilay,igpt) * f
-          tau(icol,ilay,igpt) = (1._wp - wf) * tau(icol,ilay,igpt)
-          ssa(icol,ilay,igpt) = (ssa(icol,ilay,igpt) - wf) /  max(eps,(1.0_wp - wf))
-          g  (icol,ilay,igpt) = (g  (icol,ilay,igpt) -  f) /  max(eps,(1.0_wp -  f))
-        end do
-      end do
-    end do
-
-  end subroutine delta_scale_2str_k
+  function delta_scale_2str_k(ncol, nlay, ngpt, tau, ssa, g)
+#      bind(C, name="delta_scale_2str_k")
+#    integer,                               intent(in   ) :: ncol, nlay, ngpt
+#    real(wp), dimension(ncol, nlay, ngpt), intent(inout) ::  tau, ssa, g
+
+#    real(wp) :: f, wf
+#    integer  :: icol, ilay, igpt
+
+    wp = eltype(f)
```

Comment:
```suggestion
    wp = eltype(tau)
```
Tau exists in every type of `optical_props`, so it makes sense to grab the eltype using it.